More types; more lattices

lambdageek opened the pull request:
I should probably split this up into multiple PRs

1. A few more refactoring commits to break up the permission analysis functions.
2. Switch from 'CallTree' to a 'Sequence' of 'CallTree's ie represent the straight-line sections using a separate datatype instead of a binary `Seq` constructor  (and then flip from lists to `Data.Sequence.Seq`)
3.  Refactor the function that applies a `PermissionAction` at a `Site` to emphasize that it's mostly pure code
4. (The big one) Use a `BoundedJoinSemiLattice` as a representation of `PermissionPresence` (actually a product of a 2 elt lattice and of a discrete 2 element poset with added top and bottom) and then use a finite map to represent a `PermissionPresenceSet` (see commit message and comments).

After all that, the only change I had to make to the test was to switch the order of some strings that came out sorted differently.  Success

lambdageek committed:
[Check.Permissions] Pull one more arg out of propagatePermissionsNode

lambdageek committed:
[Check.Permissions] Use the Any monoid for growing (NFC)

Maybe this reads a little better?

lambdageek committed:
[Check.Permissions] Pull processCallTree  out of propagatePermissionsNode

lambdageek committed:
Refactor CallTree into sequence of choices

Basically the canonical form we want is a sequence of calls and choices
where the branches of the choice are themselves sequences.

lambdageek committed:
Change CallSequence to use a Data.Sequence.Seq instead of a list

lambdageek committed:
[Check.Permissions] factor out permissionsPresenceFromCalleeActions

lambdageek committed:
[Check.Permissions] Refactor permissionsPresenceFromCalleeActions

lambdageek committed:
Reformat some comments to make Haddock happy (NFC)

lambdageek committed:
Add Algebra.Algebra - (bounded) join & meet semilattice typeclasses

This is cribbed from
https://hackage.haskell.org/package/lattices/docs/Algebra-Lattice.html

except my HashMap instance uses unionWith instead of union

lambdageek committed:
Use a lattice representation of PermissionPresenceSet

Represent PermissionPresence as a product lattice of Usage and Capability.

* Usage is a simple two element lattice - either we know nothing about how a
  permission name is used, or we know that it's used by a call that has a 'Uses'
  action.

* Capability is a 4 element lattice that tracks what we know about a permission
  name at each site in a call sequence: CapUnknown (bottom) - we don't know anything,
  CapHas - the site has the capability, CapLacks - the site lacks
  the capablity, CapConflict - we learned conflicting information (both
  'CapHas' and 'CapLacks') along different paths in the call tree.

* PermissionPresence is a simple product lattice: its bottom is the pair bottom
  bottom and its join and partial order are given component-wise.

The PermissionPresence lattice extends to a lattice on
PermissionPresenceSet (now a finite map from PermissionName to
PermissionPresence): the empty set is the bottom element, and the join of two
finite maps takes their union applying the join operation of PermissionPresence
when the same name is present in both maps.

With this, we can see that most of the propagation during inference is simple
joins of old and new information except when propagating forward through 'Call'
site which has a 'Grant' or 'Revoke' action which overwrites the Capability
portion of the prior information (which are in some sense stale from the
previous iteration).

lambdageek committed:
[test] fixup golden strings

lambdageek committed:
Refactor: don't compute PermissionAction -> PermissionPresence multiple ways

both the initial construction of PermissionPresence from PermissionAction, and
the computation of pre- and post-conditions when working at a 'Call' site can
be decomposed in the same way

lambdageek commented:
@evincarofautumn any thoughts about this?  In particular does the lattice representation jibe with your idea of how the Ward is designed?

lambdageek committed:
nodeAnnotations don't need to be mutable

lambdageek commented:
I have a follow up to this  (https://github.com/lambdageek/Ward/tree/hack-fewer-IO) that makes a lot more of the code pure (and hopefully QuickCheck-able) but that temporarily lost the ability to check restrictions at intermediate nodes.  I ran out of time this week trying to bring that part back.  

Anyway, I hope we can take this one because things get a lot prettier from here out.

evincarofautumn commented:
Yeah, this looks reasonable. Grant & revoke actions are lateral changes in the lattice—which I guess are still monotonic after all, hah. The distinction between use & need is still a little strange when presented like this, although admittedly it was strange already. I think we need to have a chat to hash out the details. As I said, it *seems* to work, but I’d rather take the time to find a formulation we’re confident about than find out later that it isn’t really correct or useful.

lambdageek commented:
For me personally, use/need actually makes more sense this way: use is kind of an orthogonal flag that indicates intent - it's there for the convenience of an annotation writer.

The thing that I'm still not clear on (but which is better on my `hack-fewer-IO` branch) is whether this is a forward or a backward dataflow analysis.  As presently formulated it's "both" which is not something you usually see.  But! I think at least some of the backward steps actually commute with the forward steps (because they are all just lattice joins and joins are commutative).  The only one that isn't a join is `strongUpdateCap` which is a meet and a join and `permissionDiff` which is a weird set difference.  But I think that one can probably be made to commute by some application of the distributive property for our lattice.

matthargett commented:
In the static analysis product I wrote (BugScan, which shipped on mono/Linux in 2003), we did a "hybrid" approach of both bottom-up (backward) and top-down (forward) data flow with great effect. Some people prefer a "pure" approach, but we found pragmatism and a focus on finding bugs over conceptual purity produced better results :)

Very excited for these changes and the next batch!

lambdageek commented:
Rebased and changed base branch to `master`.

@matthargett Thanks for dropping by - always happy for input.  If you have suggestions for us, we'd be glad to hear them - do make some issues!

bgamari commented:
What is the status of this?

lambdageek commented:
I got distracted and didn't work on it.  Also I don't remember if @evincarofautumn agreed with the general direction that I was going with this.

I think this PR (except the merge conflicts) was good to go in terms of passing tests. The [hack-fewer-IO branch](https://github.com/lambdageek/Ward/tree/hack-fewer-IO) was also good.

lambdageek commented:
force-pushed to fix conflicts.

@evincarofautumn take a look at it one more time?

bgamari commented:
For what it's worth I have rebased this work on top of my own branch (https://github.com/bgamari/Ward) which contains a number of significant performance improvements. I can now verify the GHC runtime system in about 10 seconds.

evincarofautumn commented:
@bgamari Those changes look good! I’m down to upstream them in addition to this.

@lambdageek Is Mono is still using this? I can do a more formal release tagging if the version isn’t already pinned and it would break anything downstream.

evincarofautumn commented:
Included in #36
